### ISSUE TITLE
Disabled cookies gives misleading "Browser can't run element" error

### ISSUE THREAD

**Issue by 0cjs** [vector-im/riot-web#14533]
### Description

Loading app.element.io in a browser with cookies disabled gives a misleading error message about the browser not being able to run Element, when in reality the issue may be fixed simply by enabling cookies.

(I am not familiar with the frequency with which people browse with cookies disabled, but doing so is a minimally invasive process, at least with Chrome where the interface allows you to persistently enable cookies for a site with three quick mouse clicks within an existing tab.)

### Steps to reproduce

- Start Chrome. (Version/platform information is below).
- From the Settings page, under "Privacy and Security," click "Cookies and other site data" and select the "Block all cookies" setting. (This is my default setting.) Use the search box at the upper right (not the centred search box above that) to search for "element" and ensure that app.element.io or a pattern matching it does not appear in the list of allowed sites.
- Open a new tab and navigate to https://app.element.io/ . The cookie icon at the right of the URL bar will have a red "X" in the lower corner. You will see a message saying "Unsupported browser," "Your browser can't run Element," "Element uses many advanced browser features, some of which are not available or experimental in your current browser." ![image](https://user-images.githubusercontent.com/2586652/87657936-762e7780-c796-11ea-9741-827d2f78734f.png)
- Click the cookie icon at the right of the URL bar; in the small pop-up dialogue box select "Always allow app.element.io to set cookies," and click the button to reload the page.  The error message will be replaced by the expected, working app.

### Version information

- **Platform**: Web
- **Browser**: 83.0.4103.116 (Official Build) (64-bit) (cohort: Stable)
- **OS**: " Windows 10 Pro version 1909
- **URL**: https://app.element.io/

**Comment by t3chguy** [comment]
> when in reality the issue may be fixed simply by enabling cookies.

There is no way to tell from a code perspective whether your browser had the capability of local storage ripped our or merely disabled.

Cookies actually aren't needed, but most "Cookie disabling" methods actually prevent IndexedDB or Local Storage from working, those are necessary for a client-side app to function. That's where your credentials and keys get stored.

**Comment by 0cjs** [comment]
Still, wouldn't explaining a bit more about what the code had tried to do and failed improve the error message? Right now in this situation it's rather misleading, especially since the suggested solution ("install Chrome") is clearly incorrect in this case.

**Comment by t3chguy** [comment]
Design clearly outlined to not specify the internals of the error but will pass it back to them again

**Comment by 0cjs** [comment]
I see. Could you point me at wherever they documented the rationale for that design decision? Perhaps I'm missing something here.

**Comment by t3chguy** [comment]
They didn't want to have to maintain the vast number of errors that are possible at this time

**Comment by 0cjs** [comment]
And those were the only two approaches they considered? Maintain a "vast" number of error messages or give an incorrect error message in some circumstances?

Anyway, seems like you guys aren't interested in addressing this, so I'll just remember that and we can move on.

**Comment by t3chguy** [comment]
I raised it for design to review in case their opinion has changed since already.

**Comment by 0cjs** [comment]
Ok, thanks!

**Comment by niquewoodhouse** [comment]
Feels to me, just from reading this, like even just saying underneath the browser links "If you have these installed and up to date already and still can't see Element, check if your cookies are disabled. Element requires cookies to run" might be of a lot of help, and doesn't require updating a big list of things?

Or am I missing the point a bit?

**Comment by t3chguy** [comment]
> still can't see Element, check if your cookies are disabled. Element requires cookies to run

It doesn't require Cookies though but some browsers upon disabling "Cookies" disable a wide assortment of issues.

This view comes up when any of ~30 tested browser feature is missing so that would be even more misleading IMO

**Comment by niquewoodhouse** [comment]
Ok, thanks for feeding back, so is it a case of my suggested copy being clearer?

> If you have these installed and up to date already and still can't see Element, check if your cookies are disabled as it might be that your browser has made some features of Element reliant on cookies being enabled."

**Comment by t3chguy** [comment]
Sure, it can also happen for when your computer is running low on Storage or due to various extensions

the actual missing feature is either Local Storage or IndexedDB, saying we use cookies might upset privacy users

**Comment by 0cjs** [comment]
Yeah, I can see that there's probably a difference between using Local Storage and locally storing cookies that someone who knows the details about these things would distinguish. But I think it's possible to come up with a message that's reasonably clear about what's going on to those who enable and disable these sorts of things without going into massive detail about it. Perhaps something along the lines of:

> __The Element App is unable to store the local data it needs to run.__
>
> You may be using a browser that doesn't support the features needed for this, or these features may be disabled by an extension or a setting such as "Disable cookies."
>
> We suggest you use Chrome, Firefox or Safari and ensure that your browser and extension settings do not disable local storage features.

Note that I'm using the generic term "local storage" here, rather than mentioning any one particular technology. There might be a better way of phrasing that to make it clear that one's talking about an entire set of related features rather than one particular feature.

**Comment by t3chguy** [comment]
This'll need to go through @nadonomy as he explicitly stated we don't want to show error details when I asked him during building this

**Comment by 0cjs** [comment]
It would be good to know the scenarios he was envisioning where it was better to give an incorrect message than a correct one. If it was along the lines of, "We don't want to show a message that says that the `somebrowser.storage.rfc9999.frob` call failed with return code 0x173," that's perfectly understandable. But the current situation goes quite a ways beyond just avoiding being overly specific in that way.

**Comment by t3chguy** [comment]
The various possible states are listed here: https://github.com/vector-im/riot-web/blob/57d2026a4007a3de8aec930988a13515598d3d7d/.modernizr.json#L8-L36

**Comment by wblessen** [comment]
Coming here 5 years later, because Element gives that message still.

I am a developer and I search for a reason why it does not run on Chrome,
any normal user will not do this.

I think the 2 suggested hints would have saved us a lot of time ...

**Comment by 0cjs** [comment]
> Also what if there are multiple errors, just saying oh its because Cookies will then let the user resolve that issue only to be met with others which may or may not have useful descriptions....

And what's wrong with that? Given the choices of "one useful description for a local configuration error that a user can quickly and easily fix" versus "no useful descriptions," the former seems better to me.
